### PR TITLE
Additional tests for cid

### PR DESCRIPTION
**Pull request by boaks**
Additional tests for cid.
Introduce cid into demo-apps.

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

**Comment by boaks** [comment]
It would great, if this PR could be merged this week.

**Comment by boaks** [comment]
> About adding DTLS_CONNECTION_ID_LENGTH, I didn't see where we use it except in test/demos ?

> In more general view, I'm a bit lost about what should go in NetworkConfig and what should not ? (and so what should go in DtlsConnectorConfig ?).
> I feel that NetworkConfig should contains only CoAP concept ?
> In this case, I can not see the link. I agree there is other option which does not respect this "idea".

The pain with the current NetworkConfig seems, that it doesn't really help to configure connectors.
The configuration must then be transferred in some code. But replacing that would require more "thoughts", that I'm currently willing to spent on. As fast solution, I would only see to move it into the "element-connector". But that seems to be also not that good.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/InMemoryConnectionStore.java [hunk 1]
```diff
@@ -80,20 +80,20 @@
  * Storing and reading to/from the store is thread safe.
  * </p>
  */
-public final class InMemoryConnectionStore implements ResumptionSupportingConnectionStore {
+public class InMemoryConnectionStore implements ResumptionSupportingConnectionStore {
 
-	private static final Logger LOG = LoggerFactory.getLogger(InMemoryConnectionStore.class.getName());
+	protected static final Logger LOG = LoggerFactory.getLogger(InMemoryConnectionStore.class.getName());
```
[the remark this answers is not in the document]
Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/InMemoryConnectionStore.java]: OK

File: demo-apps/cf-nat/src/main/java/org/eclipse/californium/examples/NatUtil.java [hunk 2]
```diff
@@ -298,9 +310,32 @@ public int assignLocalAddress(InetSocketAddress incoming) throws SocketException
...
 	/**
-	 * Remove mapping for incoming inet address.
+	 * Exchange all local addresses of NAT entries.
 	 * 
-	 * @param incoming inet address to remove mapping
+	 * Reuse the local addresses for different incoming addresses.
+	 * 
+	 * @throws SocketException 
+	 */
+	public void exchangeLocalAddresses() {
```
[the remark this answers is not in the document]
Comment (boaks) [reply demo-apps/cf-nat/src/main/java/org/eclipse/californium/examples/NatUtil.java]: OK

File: demo-apps/cf-nat/src/main/java/org/eclipse/californium/examples/NatUtil.java [hunk 3]
```diff
@@ -298,9 +310,32 @@ public int assignLocalAddress(InetSocketAddress incoming) throws SocketException
...
+		Random random = new Random();
+		List<NatEntry> destinations = new ArrayList<NatEntry>();
+		Set<InetSocketAddress> keys = new HashSet<InetSocketAddress>(nats.keySet());
+		for (InetSocketAddress incoming : keys) {
+			NatEntry entry = nats.remove(incoming);
+			destinations.add(entry);
+		}
```
[the remark this answers is not in the document]
Comment (boaks) [reply demo-apps/cf-nat/src/main/java/org/eclipse/californium/examples/NatUtil.java]: It doesn't look easier to me :-).

**Comment by boaks** [comment]
Should we merge the first 3 commits?

But, as written above, I don't see a "good & fast" solution for configuring connectors.

File: demo-apps/cf-nat/src/main/java/org/eclipse/californium/examples/NatUtil.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sbernard31) [reply demo-apps/cf-nat/src/main/java/org/eclipse/californium/examples/NatUtil.java]: This is when I realize that readability is very very subjective :grimacing:

A "cheap argument" could be that `nats.clears()` is probably more efficient than remove key 1 by 1.

But I don't think we need performance here...

**Comment by sbernard31** [comment]
> Should we merge the first 3 commits?
You can. (if you still don't feel `nats.clear` is better  :stuck_out_tongue: )

> But replacing that would require more "thoughts"

I agree. Personally I haven't a clear idea about what we should do... I just have some concerns about it and I asked myself if adding more elements in `NetworkConfig` could make the "refactoring" more difficult.

>  I don't see a "good & fast" solution for configuring connectors.

Me too ! And we won't probably go for a massive refactoring about that for the 2.0.0. 

About DTLS_CONNECTION_ID_LENGTH,  it is already configurable in `DtlsConnectorConfig`, but if you feel this is necessary to add it to `NetworkConfig`, let's go for it.

**Pull request merged**